src/ripple/core/impl/Config.cpp

```diff
@@ -362,7 +362,50 @@ Config::loadFromString(std::string const& fileContents)
         PEER_PRIVATE = beast::lexicalCastThrow<bool>(strTemp);
 
     if (getSingleSection(secConfig, SECTION_PEERS_MAX, strTemp, j_))
+    {
         PEERS_MAX = beast::lexicalCastThrow<std::size_t>(strTemp);
+        legacyPeersMax_ = true;
+    }
+    else
+    {
+        std::optional<int> peers_in_max{};
```

Comment:
At least need `peers_in_max`. For instance, `peers_max` is not present and only `peers_in_max` is present and is `0`. The intention is not clear in this case - should be assume that this is the legacy path or use `peers_out_max` default value? Checking for `0` doesn't work in this case. If on the other hand `peers_out_max` is present and has a valid value and `peers_in_max` is not present then we can safely assume that it's `0`. So can remove the latter check in this case.

Reply:
Oh that's annoying. I didn't think about a "0" value in the file...

> Also could just have a bool flag instead of optional.

No, `optional`s are better than flags in this case.

Reply:
Ok, only `peers_in_max` optional, right? No need really for `peers_out_max`, unless we want to enforce that both parameters are configured.

Reply:
Actually, why is `optional` better in this case? We want to know if the section is present. If we use the optional's then we assign its value to `PEERS_IN_MAX` anyways and then use the optional as the flag. Can simply assign the parsed value to `PEERS_IN_MAX` and set the flag to `true`.

Reply:
> Ok, only peers_in_max optional, right? No need really for peers_out_max, unless we want to enforce that both parameters are configured.

I think it makes sense to keep both `optional`. It seems like these go together, and if one is configured, both should be configured, even if one of them is 0 (and the code currently enforces that). Otherwise, IMHO, it makes more sense to treat a "0" as if it was not set, and define the case where both are 0 as "use the legacy logic". By implication, we'd probably have to make it legal to only have one defined / non-zero.

I'm thinking again of where you said the user's intention is not clear if they specify "0" - instead of trying to guess what the user wants, we can legitimately define rippled's behavior, and let the user decide what they want to do within those bounds.

> Actually, why is optional better in this case?

It keeps the flag and the value together, and reminds you to check the flag before accessing the value (you could skip the check, but that would raise red flags for a reviewer), both of which reduce the risk of programming errors.

Reply:
Ok. So I'll keep both `optional` and will allow for `peers_in_max` to be 0. I think this is a legitimate value for `in` since we enforce both `in` and `out` to be defined rather then either one. The `out` has to be >= 10.